### ISSUE TITLE
Request for project inclusion: scikit-dimension

### ISSUE THREAD

**Issue by j-bac** [scikit-learn-contrib/scikit-learn-contrib#49]
Request for project inclusion in scikit-learn-contrib

* Project name: scikit-dimension: Intrinsic dimension estimation in Python
* Project description: 
Intrinsic dimension is the minimal number of variables needed to represent a dataset. Global estimators return one dimension for the entire dataset. Some datasets have regions with different dimensions; in this case one can use local estimators to return several dimension estimates, corresponding to different local regions of a dataset.

* Authors: Jonathan Bac & Andrei Zinovyev
* Current repository: https://github.com/j-bac/scikit-dimension
* Requirements:
 * [x] scikit-learn compatible (check_estimator passed)
 * [x] Documentation (guide, API reference, example gallery)
 * [x] Unit tests (coverage: 84%)
 * [x] Python3 compatible
 * [x] PEP8 compliant (formatted with black)
 * [x] Continuous integration

**Comment by adrinjalali** [comment]
Thanks for you interest. The estimators don't seem to have a `transform` method though.

**Comment by j-bac** [comment]
Thanks for this remark. The API and docs have been improved and instead of `transform` methods estimators now have more appropriate `predict`-type methods inherited from a base class. We are happy to address other remarks, let us now if the project is a good fit for scikit-learn-contrib

**Comment by adrinjalali** [comment]
any reason you think `predict` is more appropriate than `transform` here? Dimensionality reduction is mostly seen as a _transformer_ (for instance PCA), and for these transformers to be in the middle of a `Pipeline`, they need to implement the `transform` method. `predict` is only used for predictors (classifiers and regressors) usually.

**Comment by j-bac** [comment]
I referred to [this page](https://scikit-learn.org/stable/developers/develop.html#apis-of-scikit-learn-objects). Transformers like PCA return a modified version of input data, like in the example: `new_data = transformer.fit_transform(data)`. It seemed to me that my usage was closer to the example for predictors: `probability = predictor.predict_proba(data)`.

The package does not perform dimensionality reduction - given a dataset, it simply returns a scalar value or a vector of predictions for each datapoint.

But I am happy to change the API according to what you think is best

**Comment by adrinjalali** [comment]
I guess then the question I'd ask, is what is the user supposed to do with these outputs? Are your estimators something which go in the middle of a pipeline, or they should be used at the end of a pipeline?

`predict` is usually used for transformers where the output is supposed to be compared with a ground truth, which doesn't seem to be available in your case. Or are these methods more similar to a clustering method?

Your documentation says

>  The notion of intrinsic dimension (ID) intuitively refers to the minimal number of features needed to represent a dataset with little information loss.

and that doesn't seem like a predictor to me.

**Comment by j-bac** [comment]
I see two main usages for the outputs, mostly at the beginning/middle of a pipeline :

- global ID estimates can be used before dimensionality reduction in a pipeline, in order to choose how many dimensions to retain
- local ID estimates, as you say, have a goal more similar to clustering/partitioning of data. It can be useful to detect regions with different dimensionality/complexity in a dataset.
- knowledge of global and local ID is useful before applying other machine learning algorithms. High ID is a sign that the dataset (or a local region of the dataset) can suffer from the curse of dimensionality or hubness. This can inform the choice of distance metric and/or algorithm. For example one could apply [hubness reduction before nearest neighbor search 
](https://scikit-hubness.readthedocs.io/en/latest/documentation/concepts.html)

It is true that `predict`' is usually used a bit differently, and there is no ground truth here. I guess I would need an `estimate` method but it is not part of standard scikit-learn templates, so I went with `predict`

**Comment by adrinjalali** [comment]
I think then `transform` would be actually a good candidate then. One can create a `ColumnTransformer` and add your method there, the output of which can inform other estimators in a pipeline.

**Comment by j-bac** [comment]
Thanks for this advice! I have released a new version changing predict back to transform

**Comment by adrinjalali** [comment]
Sorry for the very late reply.

Checking your API, I noticed you have quite a few `fit` parameters which in sklearn they're always an `__init__` param, an example: `smooth=False, n_neighbors=None, comb='mean', n_jobs=1`.

Otherwise I think it's quite good and we can move it to the contrib org once you fix those.

**Comment by adrinjalali** [comment]
I also note that there hasn't been much activity on the repo in the past year. Is it still maintained?

**Comment by j-bac** [comment]
Thanks for the comments! 

Indeed those are parameters shared by all estimators and only useful when calling `.fit_pw()` rather than standard sklearn `.fit()`

My reasoning is that it is confusing to put them as parameters of the class directly since:
- many estimators already have an `n_neighbors` parameter (with different purpose)
- parameters  of `.fit_pw()` are not used and don't change results when simply calling `.fit()`. `.fit_pw()` is just a helper method specific to the package, which runs the fitting process in small neighborhoods

With that said, if necessary for sklearn API I'll make the change. The package is still maintained, there simply haven't been many issues opened (but each of them was addressed). New estimators should be added by end of the year and I'm very happy to integrate the package into sklearn-contrib

**Comment by adrinjalali** [comment]
the `fit_pw` isn't a part of the sklearn API anyway, so you can do as you wish there, but it'd be nice to have the others follow the sklearn norms.

I'm happy to accept the project under the org anyhow, it seems mature enough.

**Comment by j-bac** [comment]
I agree, I'll think how to adapt this for `fit`. Happy to move the repo under the org if you send the request (or do I have to send it?)

**Comment by adrinjalali** [comment]
I think you need to send a request and I can accept it.

**Comment by j-bac** [comment]
I get message "You don’t have the permission to create public repositories on scikit-learn-contrib" when trying to transfer ownership to scikit-learn-contrib

**Comment by adrinjalali** [comment]
I don't think you've given me enough permissions to move the repo.

**Comment by j-bac** [comment]
you mean transferring it to you ? I sent an invitation to do that. Will people be properly redirected from [https://github.com/j-bac/scikit-dimension/](https://github.com/j-bac/scikit-dimension/) to scikit-learn-contrib if we do this, though ? We provide this link in the paper

**Comment by adrinjalali** [comment]
I got the invitation and accepted, but I still don't see "settings" when I go to your repo. You probably need to raise my permission level.

> We provide this link in the paper

If it doesn't you can always have a fork of the repo under your account anyway and people will see that. You can also contact github after the move to change the "forked from" setting of your repo to the contrib one.

**Comment by adrinjalali** [comment]
Got it, will move it to the org.

**Comment by chkoar** [comment]
This issue should be close. The requested project is already hosted in scikit-learn-contrib.
